drivers/adodb-pdo.inc.php

```diff
@@ -64,79 +64,97 @@ function adodb_pdo_type($t)
+{
+	
 	var $databaseType = "pdo";
 	var $dataProvider = "pdo";
 	var $fmtDate = "'Y-m-d'";
 	var $fmtTimeStamp = "'Y-m-d, h:i:sA'";
 	var $replaceQuote = "''"; // string to use to replace quotes
 	var $hasAffectedRows = true;
 	var $_bindInputArray = true;
-	var $_genIDSQL;
-	var $_genSeqSQL = "create table %s (id integer)";
-	var $_dropSeqSQL;
+
+	/*
+	* Sequence management statements
+	*/
+	public $_genIDSQL 		 = '';
+	public $_genSeqSQL 	 	 = 'CREATE TABLE %s (id integer)';
+	public $_genSeqCountSQL  = '';
+	public $_genSeq2SQL 	 = '';
+	public $_dropSeqSQL 	 = 'DROP TABLE %s';
+
+	
+
 	var $_autocommit = true;
 	var $_lastAffectedRows = 0;
 
 	var $_errormsg = false;
 	var $_errorno = false;
 
 	var $stmt = false;
-	var $_driver;
+
+	/*
+	 * Holds the current database name
+	 */
+	protected $databaseName = '';
 
 	/*
 	* Describe parameters passed directly to the PDO driver
 	*
-	* @example $db->pdoOptions = [\PDO::ATTR_ERRMODE => \PDO::ERRMODE_EXCEPTION];
+	* @example $db->pdoParameters = [\PDO::ATTR_ERRMODE => \PDO::ERRMODE_EXCEPTION];
 	*/
 	public $pdoParameters = array();
 
-	function _UpdatePDO()
-	{
-		$d = $this->_driver;
-		$this->fmtDate = $d->fmtDate;
-		$this->fmtTimeStamp = $d->fmtTimeStamp;
-		$this->replaceQuote = $d->replaceQuote;
-		$this->sysDate = $d->sysDate;
-		$this->sysTimeStamp = $d->sysTimeStamp;
-		$this->random = $d->random;
-		$this->concat_operator = $d->concat_operator;
-		$this->nameQuote = $d->nameQuote;
-		$this->arrayClass = $d->arrayClass;
-
-		$this->hasGenID = $d->hasGenID;
-		$this->_genIDSQL = $d->_genIDSQL;
-		$this->_genSeqSQL = $d->_genSeqSQL;
-		$this->_dropSeqSQL = $d->_dropSeqSQL;
-
-		$d->_init($this);
-	}
-
-	function Time()
+	/**
+	 * Connect to a database.
+	 *
+	 * @param string|null $argDSN 		 	The host to connect to.
+	 * @param string|null $argUsername 	 	The username to connect as.
+	 * @param string|null $argPassword 		The password to connect with.
+	 * @param string|null $argDatabasename 	The name of the database to start in when connected.
+	 * @param bool $persist (Optional) 		Whether or not to use a persistent connection.
+	 *
+	 * @return bool|null True if connected successfully, false if connection failed, or null if the mysqli extension
+	 * isn't currently loaded.
+	 */
+	public function _connect($argDSN, $argUsername, $argPassword, $argDatabasename, $persist=false)
 	{
-		if (!empty($this->_driver->_hasdual)) {
-			$sql = "select $this->sysTimeStamp from dual";
-		}
-		else {
-			$sql = "select $this->sysTimeStamp";
+		
+		$driverClassArray = explode('_',get_class($this));
+		/*
+		* We have already instantiated the correct driver class using pdo\<driver> so the array looks like
+		*
+		Array
+		(
+			[0] => ADODB
+			[1] => pdo
+			[2] => mysql
+		)
+		* So its easy to determine that the driver we are using is mysql
+		*/
+		$driverClass      = array_pop($driverClassArray);
+	
+		$at = strpos($argDSN,':');
+		if ($at > 0)
+		{
+		
+			$this->dsnType = substr($argDSN,0,$at);
```

Comment:
Probably better to use explode() - more efficient than strpos+substr, and avoids error in case DSN begins with a `:`
```php
$dsnSplit = explode(':', $argDSN, 2);
if (count($dsnSplit) > 1) {
    $this->dsnType;
```